test/CAPI/firrtl.c

```diff
@@ -0,0 +1,231 @@
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("annotations")),
+          emptyArrayAttr),
+  };
+  BuildRegionBlock circuitBuildBlock = {
+      .numArgs = 0,
+      .args = NULL,
+      .locs = NULL,
+  };
+  BuildRegion circuitBuildRegions[] = {{
+      .numBlocks = 1,
+      .blocks = &circuitBuildBlock,
+  }};
+  Op firCircuit =
+      buildOp(ctx, mlirModuleGetBody(module), "firrtl.circuit",
+              ARRAY_SIZE(circuitAttrs), circuitAttrs, 0, NULL, 0, NULL,
+              ARRAY_SIZE(circuitBuildRegions), circuitBuildRegions, unkLoc);
+
+  FIRRTLPortDir portsDirs[] = {
+      FIRRTL_PORT_DIR_INPUT,
+  };
+  MlirAttribute portsNames[] = {
+      mlirStringAttrGet(ctx, mlirStringRefCreateFromCString("portIn")),
+  };
+  MlirAttribute portTypes[] = {
+      mlirTypeAttrGet(firrtlTypeGetUInt(ctx, 32)),
+  };
+  MlirAttribute portAnnotations[] = {
+      emptyArrayAttr,
+  };
+  MlirAttribute symAttrs[] = {};
+  MlirAttribute portLocations[] = {
+      mlirLocationGetAttribute(unkLoc),
+  };
+
+  MlirNamedAttribute moduleAttrs[] = {
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("sym_name")),
+          mlirStringAttrGet(ctx, mlirStringRefCreateFromCString("Circuit"))),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("annotations")),
+          emptyArrayAttr),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx,
+                            mlirStringRefCreateFromCString("portDirections")),
+          firrtlAttrGetPortDirs(ctx, ARRAY_SIZE(portsDirs), portsDirs)),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("portNames")),
+          mlirArrayAttrGet(ctx, ARRAY_SIZE(portsNames), portsNames)),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("portTypes")),
+          mlirArrayAttrGet(ctx, ARRAY_SIZE(portTypes), portTypes)),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx,
+                            mlirStringRefCreateFromCString("portAnnotations")),
+          mlirArrayAttrGet(ctx, ARRAY_SIZE(portAnnotations), portAnnotations)),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("portSyms")),
+          mlirArrayAttrGet(ctx, ARRAY_SIZE(symAttrs), symAttrs)),
+      mlirNamedAttributeGet(
+          mlirIdentifierGet(ctx,
+                            mlirStringRefCreateFromCString("portLocations")),
+          mlirArrayAttrGet(ctx, ARRAY_SIZE(portLocations), portLocations)),
+  };
+  BuildRegionBlock moduleBuildBlock = {
+      .numArgs = 0,
+      .args = NULL,
+      .locs = NULL,
+  };
+  BuildRegion moduleBuildRegions[] = {{
+      .numBlocks = 1,
+      .blocks = &moduleBuildBlock,
+  }};
+  Op firModule =
+      buildOp(ctx, firCircuit.regions[0].blocks[0], "firrtl.module",
```

Comment:
There is a lot of code in this test to build IR, but building IR isn't really what we're testing here. I think it is simpler to just parse textual IR from a string, for example: https://github.com/llvm/circt/blob/68cfcd1a43f290e63cf979eb7e48bfcb0d5df55b/test/CAPI/om.c#L21-L36

Reply:
`buildOp` is a very general function, which can be used to build all other firrtl operations in case of we adding other tests in the future.

Another reason why I wanted to keep it there is that the code can be a sort of demo code for new users who don't know about these things to get started.

Reply:
My perspective is this is a test for the exporter, and it should be focused on the minimal thing necessary to test the exporter. The `buildOp` is also testing all of the C APIs for constructing IR. That's great, but I think they should be a separate test to test those C APIs specifically. I think a test for the exporter should just parse the IR and make sure it can be exported.

Reply:
Generally, if we can keep the tests shorter and more maintainable, that's preferable over making them demo code.